File: app/src/main/java/yagoc/board/Square.java
package yagoc.board;

import yagoc.pieces.PieceColor;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

public class Square implements Serializable, Comparable<Square> {
    public static List<Square> allSquares = all();

    private final int rank;
    private final int file;

    private static List<Square> all() {
        List<Square> squares = new ArrayList<>();

        for (int rank = 0; rank <= 7; rank++) {
            for (int file = 0; file <= 7; file++) {
                squares.add(new Square(rank, file));
            }
        }

        return squares;
    }

    public Square nextRank(PieceColor pieceColor) {
        if (pieceColor == PieceColor.whiteSet) {
            return new Square(rank - 1, file);
        } else {
            return new Square(rank + 1, file);
        }
    }

    public Square next2Rank(PieceColor pieceColor) {
        if (pieceColor == PieceColor.whiteSet) {
            return new Square(rank - 2, file);
        } else {
            return new Square(rank + 2, file);
        }
    }

    public Square next2File(PieceColor pieceColor) {
        if (pieceColor == PieceColor.whiteSet) {
            return new Square(rank, file + 2);
        } else {
            return new Square(rank, file - 2);
        }
    }

    public Square previous2File(PieceColor pieceColor) {
        if (pieceColor == PieceColor.whiteSet) {
            return new Square(rank, file - 2);
        } else {
            return new Square(rank, file + 2);
        }
    }

    public Square nextRankNextFile(PieceColor pieceColor) {
        if (pieceColor == PieceColor.whiteSet) {
            return new Square(rank - 1, file + 1);
        } else {
            return new Square(rank + 1, file - 1);
        }
    }

    public Square nextRankPreviousFile(PieceColor pieceColor) {
        if (pieceColor == PieceColor.whiteSet) {
            return new Square(rank - 1, file - 1);
        } else {
            return new Square(rank + 1, file + 1);
        }
    }

    public Square previousFile(PieceColor pieceColor) {
        if (pieceColor == PieceColor.whiteSet) {
            return new Square(rank, file - 1);
        } else {
            return new Square(rank, file + 1);
        }
    }

    public Square nextFile(PieceColor pieceColor) {
        if (pieceColor == PieceColor.whiteSet) {
            return new Square(rank, file + 1);
        } else {
            return new Square(rank, file - 1);
        }
    }

    public Square previousRank(PieceColor pieceColor) {
        if (pieceColor == PieceColor.whiteSet) {
            return new Square(rank + 1, file);
        } else {
            return new Square(rank - 1, file);
        }
    }

    public Square previous2Rank(PieceColor pieceColor) {
        if (pieceColor == PieceColor.whiteSet) {
            return new Square(rank + 2, file);
        } else {
            return new Square(rank - 2, file);
        }
    }

    public boolean exists() {
        return this.rank <= 7 && this.rank >= 0 && this.file <= 7 && this.file >= 0;
    }

    public List<Square> diagonalSquares() {
        List<Square> acc = new ArrayList<>();
        for (int distance = 1; distance <= 7; distance++) {
            acc.add(new Square(rank + distance, file + distance));
            acc.add(new Square(rank + distance, file - distance));
            acc.add(new Square(rank - distance, file + distance));
            acc.add(new Square(rank - distance, file - distance));
        }
        return acc.stream().filter((Square::exists)).collect(Collectors.toList());
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Square square = (Square) o;
        return rank == square.rank &&
                file == square.file;
    }

    @Override
    public int hashCode() {
        return Objects.hash(rank, file);
    }

    public List<Square> straightSquares() {
        List<Square> acc = new ArrayList<>();
        for (int distance = 1; distance <= 7; distance++) {
            acc.add(new Square(rank + distance, file));
            acc.add(new Square(rank - distance, file));
            acc.add(new Square(rank, file + distance));
            acc.add(new Square(rank, file - distance));
        }
        return acc.stream().filter((Square::exists)).collect(Collectors.toList());
    }

    @Override
    public String toString() {
        return "(" + Move.RANK_NAMES[rank] + ", " + Move.FILE_NAMES[file] + ')';
    }

    public int rank() {
        return rank;
    }

    public int file() {
        return file;
    }


    public Square(int rank, int file) {
        this.rank = rank;
        this.file = file;
    }

    /**
     * position of the square in a sequence from 0 to 63
     */
    public Integer arrayPosition() {
        return rank * 8 + file;
    }

    public Squares toSquares() {
        return Squares.Companion.valueOf(arrayPosition());
    }

    @Override
    public int compareTo(Square o) {
        return this.arrayPosition().compareTo(o.arrayPosition());
    }
}

File: app/src/main/kotlin/yagoc/board/Square.kt
package yagoc.board

import yagoc.pieces.PieceColor
import java.io.Serializable
import java.util.*
import java.util.stream.Collectors

data class Square(val rank: Int, val file: Int) : Serializable, Comparable<Square> {
    constructor(arrayPosition: Int) : this(arrayPosition / 8, arrayPosition % 8)

    fun rank() = rank

    fun file() = file

    fun nextRank(pieceColor: PieceColor): Square {
        return if (pieceColor == PieceColor.whiteSet) {
            Square(rank - 1, file)
        } else {
            Square(rank + 1, file)
        }
    }

    fun next2Rank(pieceColor: PieceColor): Square {
        return if (pieceColor == PieceColor.whiteSet) {
            Square(rank - 2, file)
        } else {
            Square(rank + 2, file)
        }
    }

    fun next2File(pieceColor: PieceColor): Square {
        return if (pieceColor == PieceColor.whiteSet) {
            Square(rank, file + 2)
        } else {
            Square(rank, file - 2)
        }
    }

    fun previous2File(pieceColor: PieceColor): Square {
        return if (pieceColor == PieceColor.whiteSet) {
            Square(rank, file - 2)
        } else {
            Square(rank, file + 2)
        }
    }

    fun nextRankNextFile(pieceColor: PieceColor): Square {
        return if (pieceColor == PieceColor.whiteSet) {
            Square(rank - 1, file + 1)
        } else {
            Square(rank + 1, file - 1)
        }
    }

    fun nextRankPreviousFile(pieceColor: PieceColor): Square {
        return if (pieceColor == PieceColor.whiteSet) {
            Square(rank - 1, file - 1)
        } else {
            Square(rank + 1, file + 1)
        }
    }

    fun previousFile(pieceColor: PieceColor): Square {
        return if (pieceColor == PieceColor.whiteSet) {
            Square(rank, file - 1)
        } else {
            Square(rank, file + 1)
        }
    }

    fun nextFile(pieceColor: PieceColor): Square {
        return if (pieceColor == PieceColor.whiteSet) {
            Square(rank, file + 1)
        } else {
            Square(rank, file - 1)
        }
    }

    fun previousRank(pieceColor: PieceColor): Square {
        return if (pieceColor == PieceColor.whiteSet) {
            Square(rank + 1, file)
        } else {
            Square(rank - 1, file)
        }
    }

    fun previous2Rank(pieceColor: PieceColor): Square {
        return if (pieceColor == PieceColor.whiteSet) {
            Square(rank + 2, file)
        } else {
            Square(rank - 2, file)
        }
    }

    fun exists(): Boolean {
        return rank in 0..7 && file in 0..7
    }

    fun diagonalSquares(): List<Square> {
        val acc: MutableList<Square> = ArrayList()
        for (distance in 1..7) {
            acc.add(Square(rank + distance, file + distance))
            acc.add(Square(rank + distance, file - distance))
            acc.add(Square(rank - distance, file + distance))
            acc.add(Square(rank - distance, file - distance))
        }
        return acc.stream().filter { obj: Square -> obj.exists() }.collect(Collectors.toList())
    }

    fun straightSquares(): List<Square> {
        val acc: MutableList<Square> = ArrayList()
        for (distance in 1..7) {
            acc.add(Square(rank + distance, file))
            acc.add(Square(rank - distance, file))
            acc.add(Square(rank, file + distance))
            acc.add(Square(rank, file - distance))
        }
        return acc.stream().filter { obj: Square -> obj.exists() }.collect(Collectors.toList())
    }

    override fun toString(): String {
        return "(" + Move.RANK_NAMES[rank] + ", " + Move.FILE_NAMES[file] + ')'
    }

    /**
     * position of the square in a sequence from 0 to 63
     */
    fun arrayPosition(): Int {
        return rank * 8 + file
    }

    override fun compareTo(other: Square): Int {
        return arrayPosition().compareTo(other.arrayPosition())
    }

    companion object {
        @JvmField
        var allSquares = all()
        private fun all(): List<Square> {
            val squares: MutableList<Square> = ArrayList()
            for (rank in 0..7) {
                for (file in 0..7) {
                    squares.add(Square(rank, file))
                }
            }
            return squares
        }
    }
}